src/lib/ares_private.h

```diff
@@ -315,6 +325,22 @@ struct ares_channeldata {
 
   /* Query Cache */
   ares__qcache_t                     *qcache;
+
+  /* Callback triggered when a server has a successful or failed response. */
+  ares_server_state_callback server_state_cb;
+  void                      *server_state_cb_data;
+
+  /* Fields controlling server failover behavior.
+   * The retry chance is the probability (1/N) by which we will retry a failed
+   * server instead of the best server when selecting a server to send queries
+   * to.
+   * The retry delay is the minimum time in milliseconds to wait between doing
+   * such retries. The next retry time tracks the timestamp when this wait
+   * expires.
+   */
+  unsigned char  server_retry_chance;
+  size_t         server_retry_delay;
+  struct timeval server_next_retry_time;
```

Comment:
`server_next_retry_time` should really be a per-server value, not per-channel as each server's retry should be timed independently.

Reply:
Sounds good. I have re-implemented the next retry time to be a field on the `server_state`.